---
title: "Analysis Report 1: Analysis of Human Skin Micorbiome in Different Age Groups"
author: "Leila Esmaeili"
date: "November 12, 2018"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction

The main purpose of this report is to explore how the data collected by Fierer et al can be used to investigate the relation between the skin microbiome and aging. It is important to understand the relationship between aging and human skin microbiome because this understanding can lay the foundation for creation of novel therapeutics that could slow down aging. This represents a paradigm shift for medicine in which our relationship with microbes is viewed as a symbiosis instead of a potential source for clinical infectious disease[@sleator2010human].


Several studies have been done in C. elegans., a model organism, to find out relationships between the microbiome and aging. For instance, it has been shown that  E. coli  that co-exist with the worms secrete diffusible molecules, including metabolites and small RNAs that can impact C. elegans aging. [@heintz2014you]  Another study has revealed how drug action on host-microbiome interactions can impact longevity in C. elegans   Metformin ( a drug widely prescribed to treat Diabetes)  extends lifespan in C. elegans  by inhibiting l-folate and methionine metabolism in E. coli. This, in turn, leads to altered methionine metabolism in the worm, and increased lifespan[@cabreiro2013metformin]. 


In addition, several studies done in humans reveal interesting information about how different age groups host different kind of Microbiome.  For example, it has been shown that human skin microbiome shifts significantly in puberty specially[@oh2012shifts].  Another study  demonstrates changes in microbiome diversity related to age[@ying2015influence].  In addition, a study has shown a certain bacterium added to the skin can help keep the skin younger [@dimarzio2008increase]).


It is worthwhile to note here the recent advances in mirobiome research has been in part made possible through novel DNA sequencing methodology.  Considering that it there are about 100 trillion bacteria associated with humans that outnumber our human cells by a factor of ten[@huttenhower2012structure],[@grice2011skin], it is important to apply these novel methods to identify bacterial sequences with accuracy and precision. A common approach used to identify bacterial populations is based on sequencing of the small subunit bacterial 16S ribosomal RNA (rRNA) gene. 16S ribosomal RNA (rRNA) gene is present in all bacteria and archaea but not in eukaryotes. The 16S rRNA gene has variable regions, used for taxonomic classification, and conserved regions, as binding sites for PCR primers. Importantly, an organism does not need to be cultured to determine its type by 16S rRNA sequencing.  Hypervariable regions within this gene contain species-specific sequences, that when compared to reference databases, allow identification of the bacteria of origin.[@lane1985rapid],[@dethlefsen2008pervasive].
 
# Methods


## Sample origin and sequencing
Fierer et al have used the following method to collect, amplify and sequence their data.   First,  they swabbed  individual keys of three personal computer keyboards and the skin on the front surface of the distal joint of each fingertip of the owner of each keyboard.  They also  swabbed space bar keys from 15 other private and public computer keyboards located on the University of Colorado campus. The  keyboards were swabbed 1–2 h after having last been touched. Skin surfaces and keyboard keys were sampled using autoclaved cotton-tipped swabs premoistened with a sterile solution All swabs were stored at −80 °C for less than 1 week before DNA extraction. 

Second, genomic DNA was extracted from the swabs using the MO BIO Power Soil DNA Isolation kit. The cotton tips of frozen swabs were broken off, placed in tubes and incubated at 65 °C for 10 min and then, for 2 minutes, shaken at maximum speed using the MO BIO vortex adapter. The remaining steps were performed as directed by the manufacturer.  Bacterial DNA was extracted from the swabs, and barcoded pyrosequencing procedure was used to determine composition.  
 
Third, for each sample, 16S rRNA genes were amplified using the primer set that had been optimized for the phylogenetic analysis of pyrosequencing reads . PCR reactions were carried out in triplicate 25-μL reactions with 0.6 μM forward and reverse primers, 3 μL template DNA, and 1× of HotMasterMix (5 PRIME). Amplicons were visualized on 0.1% agarose gels using SYBR Safe DNA gel stain in 0.5× TBE (Invitrogen). Amplicons were cleaned using the UltraClean-htp 96-well PCR Clean-up kit (MO BIO). Amplicon DNA concentrations were measured using the Quant-iT PicoGreen dsDNA reagent and kit (Invitrogen). Then, cleaned amplicons were combined in equimolar ratios into a single tube[@fierer2010forensic]. 

Lastly, The DNA was separated on ice for 45 min after the addition of 5 M NaCl (0.2 M final concentration) and 2 volumes of ice-cold 100% ethanol. The precipitated DNA was centrifuged at 7,800 × g for 40 min at 4 °C, and the resulting pellet was washed with an equal volume of ice-cold 70% ethanol and centrifuged again at7,800×g for20minat4°C. The super natant was removed and the pellet was air dried for 10 min at room temperature and then resuspended in nuclease- free water (MO BIO). Pyrosequencing was carried out on a 454 Life Science [@fierer2010forensic]. 

## Computational 

To analyze this data,  several libraries "dplyr", tidyr, "knitr", “ggplot2”, “Citr”, “seqinr”,”mctoolsr” were downloaded into R in addition to DADA2 and phyloseq.  The main job of clearning the dataset and  eliminating duplicate sequences is done through Dada2 package while Phyloseq, installed from Bioconductor, is used to visualize our result. Other libraries such as tidyr and ggplot are used to create graphs where as Citr is used to create references. Seqinr is used to export a fasta of our final denoised sequence variant. 
 
The dada2 package infers exact amplicon sequence variants (ASVs) from high-throughput amplicon sequencing data. . The dada2 pipeline takes as input demultiplexed fastq files, and outputs the sequence variants and their sample-wise abundances after removing substitution and chimera errors. Taxonomic classification is available using  the RDP naive Bayesian classifier, and species-level assignment to 16S rRNA gene fragments by exact matching [@callahan2016], [@mcmurdie2013].   
  
Some of the steps used to process the data is as follows: Initally, the files were sorted and the .Fasta ending was then eliminated.   Quality plots for all samples were then constructed. Qualtiy profiles demonstrated that most reads tend to be lower in quality after around 200 basepairs.  DADA2 was  used to trim data for maximum length of 300. Also, duplicates sequences were removed through dereplication- combining all identical sequencing reads into into “unique sequences”.  Several graphs were then created to show error rates  for each possible transition (A→C, A→G, …). On these graphs, Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here the estimated error rates (black line) are a good fit to the observed rates (points), and the error rates drop with increased quality as expected. The data looks reasonable and clean and we can proceed with confidence[@edgar2015error].  



# Results

I was curious to find out how many unique age groups are present in our data set and what gender those induvial are.  Therefore, I created Table 1. According to this table, there are 3 unique age groups in the dataset: Ages 25, 33, 36.  

To examine further the differences in terms of sequence abundance in these 3 age groups,  Figure 1 was constructed.   This figure shows that the abundance of sequences in 25  year olds is about 18,000 whereas the number is approximately 15,000 in 33 year olds and apporximately 10,000 in 25 year old age group. I also wanted to find out the abundance of different classes and phyla  in all groups combined.  Figure 2 demonstrates that in all groups combined Actinobacteria is the most abundant group with abundance of over 60,000 sequences. 


Since bacterial abundance is highest in age group 25,  I was curious to find out how much diversity is seen in this group.  I constructed  Figure 3 to find out the diversity in age group 25.  In age group 25 years old  Acientobactor is the largest class followed by Bacilli.  

In age group 33 and 36 , abundance of sequences is  shown in Figure 4 and Figure 5 respectively.   Majority of Sequences found in age group 33 years old are Actinobacteria following Baccilli, Chlororplast, Garmmaproteobactera. In age 36, Actinobacteri, Baccili and Clostriia are the top three highest counts in that order. 
```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final denoised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this -- devtools::install_github("leffj/mctoolsr")
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2017

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```
```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:20])
```


We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. 

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 300,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of final sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, denoised sequence variants
# so that we can build a phylogeny - we'll build the phylogeny outside of R
# but we need the fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is rather confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                    "fierer_hand_bacteria_SRA_study_ERP022626_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_FastTree.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants
```



```{r table-age-sex}
# make a table to count the number of samples in the intersection of
# two categorical columns
table(metadata_in$age,
      metadata_in$Sex)
```

** Table 1** This table counts the number of samples in the intersection of 
uniqe age groups and and sex.  The table indicates that there are 3 sepecific 
age groups 25, 33, and 36 year old with the gender being male.

```{r abundance-across-age-groups}
# bar plot of age group and Abundance of sequences
plot_bar(phyloseq_obj, x = "age")
```

**Figure 1** This graph is a histogram that shows Abundance of seqences across 
age groups. The largest number of  sequences was seen in age group 25 years old.  

```{r histogram-shows-diversity-all-age-groups-combined}
# subset out only the fingertip samples
age_phyloseq_obj <- subset_samples(phyloseq_obj)
                           
# bar plot of finger samples and Abundance of sequences in different
# Classes, colored by Order
plot_bar(age_phyloseq_obj,
         x = "Class",
         fill = "Order")
```

** Figure 2** Diversity of phyla on fingertips colored by class.  
This histogram shows that actinobacteria is the largest class and phylum present
on fingertips.

```{r histogram-diversity-age-25}
# Abundance of  sequenecs in age group 25
age_phyloseq_obj <- subset_samples(phyloseq_obj, age == "25")
plot_bar(age_phyloseq_obj, x = "Class")
```

** Figure 3** This graph is a histogram that descibes diversity of phyla in age 
group 25 years old. Acientobactor is the largest class followed by bacilli.

```{r histogram-diversity-age33}
# Histogram abundace age 33
age_phyloseq_obj <- subset_samples(phyloseq_obj, age == "33")
plot_bar(age_phyloseq_obj, x = "Class")
```

** Figure 4** 
This graph is a histogram that describes the abundance of sequences
found in age group 33 years old with  the majority being Actinobacteria 
following baccilli, Chlororplast, Garmmaproteobactera.


```{r histogram-diversity-age-36}
# Histogram abundance age 36
age_phyloseq_obj <- subset_samples(phyloseq_obj, age == "36")
plot_bar(age_phyloseq_obj, x = "Class")
```


** Figure 5** This graph is a histogram that describes the abundance of 
sequences found in age group 36 years old with  the majority being 
Actinobacteria, bacilli and clostridia.
# Discussion

The results mentioned above show that there are differences in diversity and abundance of microorganisms in each age group.  Figure 1 demonstrates that age 25 years old has the most abundance.  Although age group 25 shows the most abundance, it shows least diversity with only 2 peaks for  Acinetobacter and Bacilli.  On the other hand, Age 36 has the most diversity with least abundance on total sequences.  Figure 5 shows 8 bars each indicating presence of a different unique sequence.  The two most abundant sequences across all groups were Acinetobacter and Bacilli.

Although differences are seen in the skin microbiome in the 3 age groups in my analysis of Fierer et al data, considering that there were only 3 individuals in this study, the sample size is too small come up with statistically significant conclusions.  The difference seen in the diversity of the skin bacteria in these 3 individuals could be due to factors outside of age such as race, weight, diet, exercise schedules.  A larger sample size controlled for confounding factors is required for a good analysis. A future study could be designed with 30 people recruited in each unique age group. In addition, all members in each group need to be following a prescribed diet, be of the same gender and race, have been screened to be taking no medications and have similar Body Mass Index.  In addition, they to be using same soaps and shampoos and other skin products 30 days prior to start of the study. 

As I was searching literature, I found a study that has a better design for the purpose of the central question that I am trying to answer.  In this study done by Ying et al[@ying2015influence], a much larger sample size of 71 was recruited.  In addtion, there were certain exclusion criteria for who could be accepted into the study design in order to control for confounders. I would follow a similar design if I were to contstruct my own study. I would then perform ANOVA or ANOSIM to find out if the differnces seen between groups is statistically significant. 

In conclusion, my analysis was sucessful in exploring how the data collected by Fierer et al can be used to investigate the relation between the skin microbiome and aging. My analysis shows that there are differences in compostion and abundance of sequnces in each unique age group.  However, because of the small sample size, we are limited to draw conclusions.  Further studeis with larger sample size are needed to lead the way to the creation of novel therapeutics that could slow down aging.


# Sources Cited




